### PR TITLE
Zarr driver

### PR DESCRIPTION
**Pull request by rouault**
Another approach at creating a Zarr driver than the one of PR #3411 (see analysis done in https://github.com/OSGeo/gdal/pull/3411#issuecomment-831259626)

Scope is:
- read/creation/update of Zarr datasets
- using nominally the multidimensional API
- and also exposed with the classic 2D API (slicing needed for > 2D datasets)
- support hierarchical organization in groups
- support attributes, including a dedicated attribute to encode the CRS, and the XArray _ARRAY_DIMENSIONS attribute to encode the netCDF-like dimension concept
- support consolidated metadata
- support reading/writing datasets on cloud storage, .zip (but that comes for free using the VSI I/O layer)
- support Zarr V2 and in-progress V3 specifications
- support compression methods: zlib, gzip, zstd, lzma, lz4 (new optional build dependency), blosc (new optional build dependency), with an API allowing to register additional compressors
- support most data types: bool, [u]int[8/16/32/64], float[32/64], complexfloat[32/64], strings, compound data types (limited to what GDALExtendedDataType supports)
- limited support for filters with delta filter only as builtin, with an API allowing to register additional filters

## Tasklist

 - [X] Read support of arrays with multidimensional API
 - [X] Read support for compression methods, using an extendable API
 - [X] Read support for Zarr V2 arrays
 - [x] Read support for compound data types 
 - [x] Read support for non-GDAL-native types (such as int1, int8, uint8, half-float)
 - [x] Read support for Fortran block ordering
 - [x] Read support for CRS
 - [x] Read support for groups
 - [x] Read support for consolidated metadata
 - [x] Read support for XArray _ARRAY_DIMENSIONS
 - [x] Read support for Zarr V3
 - [x] Read support for filters (limited to delta as builtin)
 - [x] Read support for group attributes
 - [x] Read support for array attributes
 - [x] Read support with classic 2D API

 - [x] Creation support of ZarrV2 arrays with multidimensional API
 - [x] Write support for compression methods
 - [x] Write support for filters (limited to delta as builtin)
 - [x] Write support for group creation
 - [x] Write support for group attributes
 - [x] Write support for array attributes
 - [x] Write support for CRS
 - [x] Write support for consolidated metadata
 - [x] Write support for XArray _ARRAY_DIMENSIONS
 - [x] Write support for Zarr V3
 - [x] Write support with classic 2D API

 - [x] Documentation page

### CODE REVIEW

**Comment by rouault** [comment]
This is now feature ready.

## Zarr driver RFC

### Who asked for the driver and why?

Zarr is an emerging format, natively capable for multidimensional raster and cloud friendly. A number of parties have expressed a desire to see it handled in GDAL.
The Zarr V2 format has just been submitted as a potential OGC community standard: https://www.ogc.org/pressroom/pressreleases/3275

### Read / Write / ReadWrite ?

Read, creation and update

### Raster / Vector / Raster&Vector?

Raster, implementing classic 2D and multidimensional API

### Driver name info

  * Short name : Zarr

  * Long name : Zarr

  * Description: Zarr

### Should this be compiled in by default, optional build, or plugin?

Built-in by default, can be disabled in autoconf builds using standard mechanism (--disable-driver-zarr)

### Does this driver require other drivers be built?

No, but it strongly benefits from building GDAL against liblz4, liblzma, libzstd and libblosc to benefit from all compression methods

### Is there a standard(s) that the driver is working to implement?

There's a specification for ZarrV2 at https://zarr.readthedocs.io/en/stable/spec/v2.html, which is the one mentioned above as being submitted to OGC.
The driver also implements the in-progress Zarr V3 spec: https://zarr-specs.readthedocs.io/en/core-protocol-v3.0-dev/protocol/core/v3.0.html

### How feature complete will the driver be?

It implements almost all the specification, except:
- compound data types that use members of type array
- a few compression methods of [NumCodecs](https://numcodecs.readthedocs.io/en/stable/) aren't supported currently: ZFPY, BZ2
- only the Delta filter is supported

### Driver capabilities.

  Supports: Raster
  Supports: Multidimensional raster
  Supports: Subdatasets
  Supports: Open() - Open existing dataset.
  Supports: Create() - Create writable dataset.
  Supports: CreateMultiDimensional() - Create multidimensional dataset.
  Supports: Virtual IO - eg. /vsimem/
  Creation Datatypes: Byte Int16 UInt16 Int32 UInt32 Float32 Float64 CFloat32 CFloat64


### Support for the following features :

  * DMD_MIMETYPE: 
  
None

  * DMD_EXTENSIONS:  

None

  * DMD_CONNECTION_PREFIX

None

  * GDAL_DMD_OPENOPTIONLIST
```
<OpenOptionList>
  <Option name="USE_ZMETADATA" type="boolean" description="Whether to use consolidated metadata from .zmetadata" default="YES" />
</OpenOptionList>

```

  * GDAL_DMD_SUBDATASETS: Yes

  * GDAL_DCAP_OPEN : Yes

  * DCAP_CREATE : Yes

  * GDAL_DCAP_VIRTUALIO: Yes

### What external libraries does this driver depend on?

- Zlib (build requirement of GDAL)
- libdeflate optionaly (for faster Zlib compression/decompression)
- liblzma optionaly
- liblz4 optionaly
- libblosc-c optionaly
- libzstd optionaly

### Are there version requirements for the libraries?

The versions provided by Ubuntu 20.04 are sufficient to get the driver building and working.

### What licenses do the libraries have?

- Zlib: [Zlib license](https://metadata.ftp-master.debian.org/changelogs//main/z/zlib/zlib_1.2.11.dfsg-2_copyright)
- libdeflate: [Expat license](https://metadata.ftp-master.debian.org/changelogs//main/libd/libdeflate/libdeflate_1.7-1_copyright)
- liblz4: [BSD 2 clauses](https://metadata.ftp-master.debian.org/changelogs//main/l/lz4/lz4_1.9.3-2_copyright)
- liblzma: [LZMA license](https://metadata.ftp-master.debian.org/changelogs//main/x/xz-utils/xz-utils_5.2.5-2_copyright)
- libblosc-c [BSD 3 clauses](https://metadata.ftp-master.debian.org/changelogs//main/c/c-blosc/c-blosc_1.20.1+ds1-2_copyright)
- libzstd [BSD 3 clauses or GPL-v2](https://metadata.ftp-master.debian.org/changelogs//main/libz/libzstd/libzstd_1.4.8+dfsg-2.1_copyright)

### Does the driver require a binary SDK?

No

### Any external services required for the driver?

No

### Any external service providers required for the driver?

No

### Testing plan

This driver has an autotest script in `autotest/gdrivers` folder

### Maintenance plan?

Mostly have a look on where Zarr V3 goes on. Current status in documentation for it is indicated as experimental

### Who / what org is responsible for upkeep on this new driver?

Even Rouault

### Where to report bugs?

GDAL gitub issue tracker

### Draft help text

see `gdal/doc/source/drivers/raster/zarr.rst` in PR for details

### What are the conditions for removing the driver?

If nobody longer interested in keeping it in a buildable and working state

### Does the driver already have a draft implementation?

Yes

### Who implemented it? Or who will implement this driver?  Does an implementer(s) need to be found?

Even Rouault

### Is there funding needed?

No

### Who will review the code / PR?

TBD

**Comment by rouault** [comment]
Is anyone interested in giving this some review ? I know this is a bit overwhelming. I could offer some funding if that might help and being practical to setup.

**Comment by tbonfort** [comment]
> Is anyone interested in giving this some review ? I know this is a bit overwhelming. I could offer some funding if that might help and being practical to setup.

cc @thomascoquet ?

**Comment by sgillies** [comment]
@rouault I've been using zarr (the python package) at work and have some data for testing. I'll try it out later today.

**Comment by rouault** [comment]
I'll merge this PR next week if there is no objection or further comments

File: gdal/configure.ac [hunk 1]
```diff
@@ -1689,6 +1689,77 @@ fi
...
+    EXTRA_INCLUDES="-I$with_blosc/include $EXTRA_INCLUDES"
+  else
+    AC_MSG_ERROR([libblosc not found])
+  fi
+
+else
+    HAVE_BLOSC=no
+fi
+
+AC_SUBST(HAVE_BLOSC,$HAVE_BLOSC)
```
Comment (SpaceIm) [inline gdal/configure.ac:1726]: Too fragile, c-blosc may depend on lz4, snappy, zlib and zstd. A naive `AC_CHECK_LIB` may break due to missing dependencies in LIBS. c-blosc provides a pkg-config file: `blosc.pc`

Comment (rouault) [reply gdal/configure.ac:1726]: > Too fragile, c-blosc may depend on lz4, snappy, zlib and zstd. A naive `AC_CHECK_LIB` may break due to missing dependencies in LIBS. c-blosc provides a pkg-config file: `blosc.pc`

too fragile for static builds of c-blosc ? should work fine hopefully for dynamic builds of c-blosc. Anyway autoconf builds will be soon deprecated, so any improvements if any needed should go to the CMake builds

Comment (SpaceIm) [reply gdal/configure.ac:1726]: > too fragile for static builds of c-blosc ?

yes (in conan-center, by default all libs are built against static dependencies so we quickly see these kind of issues).

> Anyway autoconf builds will be soon deprecated, so any improvements if any needed should go to the CMake builds

Great ! I'll check that while packaging the first release using CMakeLists.

Comment (rouault) [reply gdal/configure.ac:1726]: > Great ! I'll check that while packaging the first release using CMakeLists.

early testing of CMake builds of master would be appreciated

**Pull request merged**